Add hooks on Presenter classes

djbuch opened the pull request:
<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Allow to add additionnal infos to each presenters thanks to new hooks
| Type?         | new feature
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | 
| How to test?  | These are just hooks allowing module developpers to add additionnal infos to presented objects

<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/16093)
<!-- Reviewable:end -->

djbuch committed:
Add hook to present method

Allow module developpers to add values to presented items

kpodemski commented:
What about performance of such a solution? the more we have presented objects the more hooks are being triggered, right?

djbuch commented:
If there is no module hooked to the hook, there will be just very marignal performance change, since module list is cached for each hook.

If there is one hook, then I think this is the best way for changing presented elements.

PierreRambaud commented:
This is exactly what we need for facetedsearch. ping @matthieu-rolland 
But, It could be cool to have a `before` and an `after` hook on each presenter :) Wdyt?

djbuch committed:
New appendLazyArray method

Add an appendLazyArray method to allow modules to add "Lazy" keys that are called only when needed.

And example usage could be :
```
public function hookPresentProductListing(&$params)
    {
        $product = $params['product'];
        $productListingLazyArray = &$params['productListingLazyArray'];

        $productListingLazyArray->appendLazyArray(
            [
                'is_abonnement' => [
                    [$this, 'isAbonnementProduct'],
                    $product['id_product']
                ]
            ]
        );

    }
```

PierreRambaud commented on src/Adapter/Presenter/AbstractLazyArray.php:
```diff
@@ -241,6 +263,12 @@ public function offsetGet($index)
+            } elseif ($this->arrayAccessList[$index]['type'] === 'callable') {
+                if (!isset($this->methodCacheResults[$index])) {
+                    $callable = $this->arrayAccessList[$index]['value'];
+                    $this->methodCacheResults[$index] = call_user_func($callable[0], isset($callable[1]) ? $callable[1] : null);
```
```suggestion
                    $this->methodCacheResults[$index] = call_user_func($callable[0], $callable[1] ?: null);
```

PierreRambaud reviewed (changes requested):
Hooks names must change: https://devdocs.prestashop.com/1.7/modules/concepts/hooks/#naming-scheme

djbuch commented:
Hello, I agree the hook names should change, but what to choose, action or display ? Because it is quite at the crossroads of both behaviors

PierreRambaud commented on src/Adapter/Presenter/AbstractLazyArray.php:
```diff
@@ -241,6 +263,12 @@ public function offsetGet($index)
+            } elseif ($this->arrayAccessList[$index]['type'] === 'callable') {
+                if (!isset($this->methodCacheResults[$index])) {
+                    $callable = $this->arrayAccessList[$index]['value'];
+                    $this->methodCacheResults[$index] = call_user_func($callable[0], $callable[1] ?: null);
```
My bad, wrong notation :D This one is better
```suggestion
                    $this->methodCacheResults[$index] = call_user_func($callable[0], $callable[1] ?? null);
```

eternoendless commented:
> I agree the hook names should change, but what to choose, action or display ? Because it is quite at the crossroads of both behaviors

If a hook returns HTML content, then it should be `display<...>`, whereas all other hooks should to be called `action<...>`.

djbuch committed:
Change hook names

djbuch commented:
OK thanks, I changed the names

PierreRambaud commented on src/Adapter/Presenter/AbstractLazyArray.php:
```diff
@@ -142,6 +142,28 @@ public function appendArray($array)
+    public function appendLazyArray($array)
+    {
+        foreach ($array as $key => $value) {
+            // do not override any existing method
```
I can't find where this method is used? :thinking:

djbuch replied:
This method is not used in the PR but is an addition to allow module developper to add "Lazy" values to an lazy array, like I showed it here : https://github.com/PrestaShop/PrestaShop/pull/16093/commits/9c32593c871a6d7e41eb1e87195ef4650c191467

eternoendless commented:
Looks like this PR may need a rebase, I see lots commits with a generic title "Update Somefile.php" plus a merge commit, and Travis is failing for some reason.

PierreRambaud commented:
Pull request has been rebased and squashed.
And I fix the code for sanity tests.

PierreRambaud commented:
@djbuch Could you provide us a modules to tests all hooks? :)

matks commented:
@djbuch Hi ! I'm very sorry I just found this PR and it looks like I already implemented it myself (see https://github.com/PrestaShop/PrestaShop/pull/17203 and https://github.com/PrestaShop/PrestaShop/pull/17419/). I was unaware you submitted this PR.

Sorry for this 😞

djbuch commented:
Hello @matks I saw your implementation, but it lacks some elements, that where in mine.
For example the 'cart' param in the actionPresentCart hook in order to get the cart that is being presented (not necessarily the one of the context).

What do you think of this ?

djbuch commented:
Hello @matks have you seen my last comment ?
